### ISSUE TITLE
Make the solveequihash benchmark more representative of solution latency / add another benchmark for throughput

### ISSUE THREAD

**Issue by daira** [zcash/zcash#1191]
@Vorksholk wrote at https://github.com/zcash/zcash/pull/1174#issuecomment-238038073 :

> Since this benchmark runs solvers on x threads but waits until all threads have finished before starting the next batch of solvers, it doesn't seem to be terribly representative of real-world throughput, since towards the end of the benchmark computational resources are sitting idle.
> 
> For example, running the zcbenchmark on 6 threads for two cycles performs the first cycle, and doesn't start the 2nd cycle until all threads of the first cycle have finished. On my 6-core system, I end up waiting ~10-15 seconds for the final thread to finish its work.
> 
> A 'simple' fix would be modifying the benchmark to solve a "pool" of x equihashes and allow threads to draw from the pool of equihashes on-demand, so that longer benchmarks would come closer and closer to approximating real-world performance since the only idle resources would exist when the pool is empty and some threads have finished their work.

**Comment by daira** [comment]
Or alternatively we could add a new throughout-oriented benchmark, target than changing the existing one.

**Comment by daira** [comment]
Since the existing benchmark is primarily measuring latency, which is a useful thing to measure independently of throughout, let's change it to report the distribution of timings of individual solver runs on all of the threads, rather than the timing of the "join" of all threads. (Then speed.z.cash will show the extrema and quartiles, which is what we actually want.)

**Comment by str4d** [comment]
Hmm, this changes the semantics of the published testing data. Currently, the extrema and quartiles on speed.z.cash show the variations across multiple separate runs of the benchmark. This would change the multi-threaded test to show variations across threads in a *single* run. It is implementable using the current benchmark API (by returning an array of `N` values where `N` is the number of threads instead of the number of runs), but this would mean that one of the benchmarks is inconsistent with the others both in terms of what the output of `performance-measurements.sh` is, and the visualisations on speed.z.cash.

**Comment by str4d** [comment]
If the above is indeed the behaviour we want, I can implement this.

**Comment by zookozcash** [comment]
I'm good with whatever you folks decide about this.

**Comment by daira** [comment]
I'm okay with changing the semantics of the multithreaded test; that test has only just been added so there isn't much historical data to compare with.

**Comment by daira** [comment]
Also, I was thinking that it would return an array of (number of threads) * (number of runs) values.

**Comment by str4d** [comment]
@daira so you're envisaging that the quartiles and extremas represent the spread including inter-run variation, not just the spread between threads in an individual run? That is, this benchmark would show the average time that a user would expect any mining thread to take to complete at any time if they have set `-genproclimit=N`? Doable, but it's a third semantic option, and I want to make sure we get the "right" one :stuck_out_tongue_winking_eye:

**Comment by daira** [comment]
Yes - because if you want the spread between runs you can use the single-threaded option, and if you want the spread between threads in a single run you can set the number of runs to 1.

**Comment by nathan-at-least** [comment]
+1 on distinct measurements of latency vs throughtput.

+0 on altering the semantics of a new test. Why not create new test names/histories for the new latency and throughput measurements instead?

**Comment by daira** [comment]
The consequences of the multi threaded test joining on all threads weren't considered when we implemented it (at least, I didn't consider that). We don't need to retain those semantics. I have no opinion on whether we should rename the test when changing it.

**Comment by nathan-at-least** [comment]
Because we're launching with the current Beta 1 zkSNARK parameters, let's postpone performance measurement improvements to after launch. (Note: there's some risk of missing performance, DOS, or pow parameter issues here, but we're taking on that risk and will adapt after launch.)

**Comment by daira** [comment]
#1430 was a duplicate, and has some good comments.

**Comment by daira** [comment]
This shouldn't have been closed: although #1376 improves the measurement of multithreaded latency, the results are still a bit misleading because they don't reflect the number of solutions found. We need a separate throughput benchmark.

**Comment by daira** [comment]
Also, I think we need that throughput benchmark before launch, if at all possible. I may work on it in my spare time.

**Comment by daira** [comment]
Since no-one uses the built-in miner, these benchmarks are no longer important.
